hnn_core/network.py

```diff
@@ -219,47 +286,411 @@ def __repr__(self):
+                                 )
+        drive['events'] = []
+
+        self._attach_drive(name, drive, weights_ampa, weights_nmda, location,
+                           space_constant, dispersion_time,
+                           cell_specific=False)
+
+    def _attach_drive(self, name, drive, weights_ampa, weights_nmda, location,
+                      space_constant, dispersion_time, cell_specific=True):
+        """Attach a drive to network based on connectivity information
+
+        Parameters
+        ----------
+        name : str
+            Name of drive (must be unique)
+        drive : instance of NetworkDrive
+            Collection of parameters defining the dynamics of the drive
+        weights_ampa : dict
+            Synaptic weights of AMPA receptors on each targeted cell type
+            (keys). Weights can be zero or left omitted.
+        weights_nmda : dict
+            Synaptic weights of NMDA receptors on each targeted cell type
+            (keys). Weights can be zero or left omitted.
+        location : str
+            Target location of synapses ('distal' or 'proximal')
+        space_constant : float
+            Describes lateral dispersion (from column origin) of synaptic
+            delays within the simulated column
+        dispersion_time : dict
+            Synaptic delay at the column origin, dispersed laterally as a
+            function of the space_constant
+        cell_specific : bool
+            Whether each cell has unique connection parameters (default: True)
+            or all cells have common connections to a global (single) drive.
+
+        Attached drive is stored in self.external_drives[name]
+        self.pos_dict is updated, and self._update_gid_ranges() called
+        """
+        if name in self.external_drives:
+            raise ValueError(f"Drive {name} already defined")
+        target_populations = (set(weights_ampa.keys()) |
+                              set(weights_nmda.keys()))
+        if not target_populations.issubset(set(self.cellname_list)):
+            raise ValueError('Allowed target cell types are: ',
+                             f'{self.cellname_list}')
+        # this is needed to keep the drive GIDs identical to those in HNN,
+        # e.g., 'evdist1': range(272, 542), even when no L5_basket cells
+        # are targeted (event times lists are empty). The logic in
+        # _connect_celltypes is hard-coded to use these implict gid ranges
+        if self._legacy_mode:
+            target_populations = self.cellname_list
+
+        drive['name'] = name  # for easier for-looping later
+        drive['target_types'] = target_populations  # for _connect_celltypes
+
+        weights_by_receptor = {'ampa': weights_ampa, 'nmda': weights_nmda}
+
+        drive['conn'], src_gid_ran = self._create_drive_conns(
+            target_populations, weights_by_receptor, location,
+            space_constant, dispersion_time, cell_specific=cell_specific)
```

Comment:
could we potentially not nest the connectivity object inside `drive` ? That way we could later add within network connectivity to the same dict? It could be an attribute of `Network` and you can document it there ...

Reply:
I'm currently a firm -1 on this one... I think it's by far the cleanest to have everything in `net.external_drives`, even though it becomes quite deeply nested. The drives are now totally "liberated" from the rest of the network, which should have plenty of benefits.

If we want to allow changes to connectivity, we certainly should't muck around with our own dicts, but rather use something like NeuroML! I can easily see that happening, at which time drive connectivity definition should also be reconsidered.

Reply:
I'm still against this proposal...

Reply:
sounds good, we can discuss this and leave modifications for later. Let's look at what NeuroML has to offer. I do want to point out that software implementation should not fully mirror specification langauges. Since software should leverage the goodness of object-oriented programming. But drawing some inspiration is a good idea